File: DSA Implementations/C++/binary_search_tree.cpp
#include <iostream>
#include <memory>
using namespace std;

struct TreeNode
{
    int value;
    unique_ptr<TreeNode> left;
    unique_ptr<TreeNode> right;

    TreeNode(int val) : value(val), left(nullptr), right(nullptr) {}
};

void printTabs(int numTabs)
{
    for (int i = 0; i < numTabs; i++)
    {
        cout << "\t";
    }
}

void printTreeRec(const unique_ptr<TreeNode> &root, int level)
{
    if (!root)
    {
        printTabs(level);
        cout << "---<empty>---\n";
        return;
    }
    printTabs(level);
    cout << "Value: " << root->value << "\n";

    printTabs(level);
    cout << "left\n";
    printTreeRec(root->left, level + 1);

    printTabs(level);
    cout << "right\n";
    printTreeRec(root->right, level + 1);

    printTabs(level);
    cout << "Done\n";
}

void printTree(const unique_ptr<TreeNode> &root)
{
    printTreeRec(root, 0);
}

bool insertNumber(unique_ptr<TreeNode> &root, int value)
{
    if (!root)
    {
        root = make_unique<TreeNode>(value);
        return true;
    }
    if (value == root->value)
        return false;

    if (value < root->value)
    {
        return insertNumber(root->left, value);
    }
    else
    {
        return insertNumber(root->right, value);
    }
}

bool findNumber(const unique_ptr<TreeNode> &root, int value)
{
    if (!root)
        return false;
    if (root->value == value)
        return true;

    if (value < root->value)
        return findNumber(root->left, value);
    else
        return findNumber(root->right, value);
}

int main()
{
    unique_ptr<TreeNode> root = nullptr;

    insertNumber(root, 15);
    insertNumber(root, 11);
    insertNumber(root, 24);
    insertNumber(root, 5);
    insertNumber(root, 19);
    insertNumber(root, 16);

    printTree(root);

    cout << 16 << " (" << findNumber(root, 16) << ")\n";
    cout << 15 << " (" << findNumber(root, 15) << ")\n";
    cout << 5 << " (" << findNumber(root, 5) << ")\n";
    cout << 115 << " (" << findNumber(root, 115) << ")\n";
    cout << 1 << " (" << findNumber(root, 1) << ")\n";
    cout << 7 << " (" << findNumber(root, 7) << ")\n";

    return 0;
}

File: DSA Implementations/Python/binary_search_tree.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def print_tabs(num_tabs):
    print("\t" * num_tabs, end="")


def print_tree_rec(root, level=0):
    if root is None:
        print_tabs(level)
        print("---<empty>---")
        return

    print_tabs(level)
    print(f"Value: {root.value}")

    print_tabs(level)
    print("left")
    print_tree_rec(root.left, level + 1)

    print_tabs(level)
    print("right")
    print_tree_rec(root.right, level + 1)

    print_tabs(level)
    print("Done")


def print_tree(root):
    print_tree_rec(root, 0)


def insert_number(root, value):
    if root is None:
        return TreeNode(value)

    if value == root.value:
        return root  

    if value < root.value:
        root.left = insert_number(root.left, value)
    else:
        root.right = insert_number(root.right, value)

    return root


def find_number(root, value):
    if root is None:
        return False
    if root.value == value:
        return True
    if value < root.value:
        return find_number(root.left, value)
    else:
        return find_number(root.right, value)

def main():
    root = None
    for num in [15, 11, 24, 5, 19, 16]:
        root = insert_number(root, num)

    print_tree(root)

    print(16, "(", find_number(root, 16), ")")
    print(15, "(", find_number(root, 15), ")")
    print(5, "(", find_number(root, 5), ")")
    print(115, "(", find_number(root, 115), ")")
    print(1, "(", find_number(root, 1), ")")
    print(7, "(", find_number(root, 7), ")")

if __name__ == "__main__":
    main()
